Adds Antag Challenge Coins

emmanuelbassil opened the pull request:
🆑emmanuelbassil
rscadd: Adds challenge coins in the traitor uplink. They unlock a secret compartment containing uplink items in any vendor on the ship. Item costs 50 TCs; comes with 5 coins making each worth 10 TCs.
tweak: Each compartment has an average worth of 16 TCs. They are usually related to the theme of the vendor. Higher cost items are not guaranteed to spawn, meaning you may get less TCs than your coin is worth. Or much more, depending on your luck.
tweak: Using a challenge coin or an emag on a vending machine makes it scream anti-Sol slogans for the benefit of all nearby non-traitors.
bugfix: Makes vending machine icons turn back on after being anchored in a powered area.
/🆑 

When complete, each vending machine in the game can accept this coin which would unlock a compartment with uplink items with an average value of 15-17 TCs. It will make use of randomization code to make this a bit of a crapshoot; if you're hilariously lucky you effectively triple the value of your TCs. If you're hilariously unlucky; you get less TCs than what you started with. Each vendor will have uplink items that are thematically appropriate to the vendor.

To do:
- [x] Make working antag challenge coin
- [x] Modify cigcase code to use overlays depending on whether you put in coins or cigarettes. Use code in #33688
- [x] Fix cigcase description. Add in-hand icon.
- [x] Improve sprite for coins in cigcase; make coin color change depending on material type. 
- [x] Create sprite for challenge coin
- [x] Make sure coin cuts access requirements
- [x] Sprite overlay with sparks to add on top of a machine after it gets a coin.
- [x] Pass by all vendors, giving each an average spawn chance of 16 TCs worth of antag items. More expensive items have a chance of not spawning in.
- [x] Make the machine start screaming anti-Sol slogans after it receives the coin.
- [x] Make sure new uplink item is in the pool of 'buy random item'. Make a pass to put other new uplink items there.

emmanuelbassil commented:
The coding/spriting part is ready for critique.
I just need a few days to pass by each vendor and add items/slogans.

emmanuelbassil commented:
This is ready for review!

MuckerMayhem reviewed (changes requested):
The styling when it comes to your if and for statements is inconsistent, add a space after every `if` and `for` you add for the sake of readability and consistency (ie `if ()` and `for ()`, same deal for other operators).

SierraKomodo commented on code/game/machinery/vending/_vending.dm:
```diff
@@ -192,20 +214,63 @@
-	if (istype(item, /obj/item/material/coin) && length(premium))
+	if (is_type_in_list(item, simple_coins))
+		if (!length(premium))
+			to_chat(user, SPAN_NOTICE("\The [src] does not accept the [item]."))
```
```suggestion
			to_chat(user, SPAN_WARNING("\The [src] does not accept \the [item]."))
```

SierraKomodo commented on code/game/machinery/vending/_vending.dm:
```diff
@@ -192,20 +214,63 @@
 		return TRUE
+	if (istype(item, /obj/item/material/coin/challenge/syndie))
+		if (!LAZYLEN(antag))
+			to_chat(user, SPAN_NOTICE("\The [src] does not have a secret compartment installed."))
```
```suggestion
			to_chat(user, SPAN_WARNING("\The [src] does not have a secret compartment installed."))
```

SierraKomodo commented on code/game/machinery/vending/_vending.dm:
```diff
@@ -192,20 +214,63 @@
+			to_chat(user, SPAN_NOTICE("\The [src] does not have a secret compartment installed."))
+			return TRUE
+		if (src.IsShowingAntag())
+			to_chat(user, SPAN_NOTICE("\The [src]'s secret compartment is already unlocked!"))
```
```suggestion
			to_chat(user, SPAN_WARNING("\The [src]'s secret compartment is already unlocked!"))
```

SierraKomodo commented on code/game/machinery/vending/_vending.dm:
```diff
@@ -192,20 +214,63 @@
+		if (!LAZYLEN(antag))
+			to_chat(user, SPAN_NOTICE("\The [src] does not have a secret compartment installed."))
+			return TRUE
+		if (src.IsShowingAntag())
```
```suggestion
		if (IsShowingAntag())
```

SierraKomodo commented on code/game/machinery/vending/_vending.dm:
```diff
@@ -192,20 +214,63 @@
+		if (src.IsShowingAntag())
+			to_chat(user, SPAN_NOTICE("\The [src]'s secret compartment is already unlocked!"))
+			return TRUE
+		if (!user.unEquip(item,src))
+			return FALSE
+		ProcessAntag(item, user)
```
```suggestion
		if (!user.unEquip(item,src))
            to_chat(user, SPAN_WARNING("You can't drop \the [item]."))
			return TRUE
		ProcessAntag(item, user)
```

emmanuelbassil commented:
Done! Ready again!

SierraKomodo commented on code/game/machinery/vending/_vending.dm:
```diff
@@ -192,20 +214,64 @@
+			to_chat(user, SPAN_NOTICE("You successfully pull the coin out before \the [src] could swallow it."))
+			return TRUE
+		else
+			to_chat(user, SPAN_NOTICE("You weren't able to pull the coin out fast enough, the machine ate it, string and all."))
```
```suggestion
			to_chat(user, SPAN_NOTICE("You weren't able to pull \the [item] out fast enough, \the [src] ate it, string and all."))
```

SierraKomodo commented on code/game/machinery/vending/_vending.dm:
```diff
@@ -192,20 +214,64 @@
+	if (!isnull(antagcoin.string_color))
+		if (prob(10))
+			user.put_in_hands(item)
+			to_chat(user, SPAN_NOTICE("You successfully pull the coin out before \the [src] could swallow it."))
```
```suggestion
			to_chat(user, SPAN_NOTICE("You successfully pull \the [item] out before \the [src] could swallow it."))
```

SierraKomodo commented on code/game/machinery/vending/_vending.dm:
```diff
@@ -192,20 +214,64 @@
+			qdel(item)
+			return TRUE
+	else
+		to_chat(user, SPAN_NOTICE("You hear a loud clink as \the [item] is swallowed by the [src]"))
```
```suggestion
		to_chat(user, SPAN_NOTICE("You hear a loud clink as \the [item] is swallowed by \the [src]"))
```

emmanuelbassil commented:
I deserve this for forgetting all \the \the's. I will never type that word without a forwardslash ever again.

Ready again!

Spookerton commented on code/game/machinery/_machines_base/machinery_components.dm:
```diff
@@ -241,6 +241,10 @@ GLOBAL_LIST_INIT(machine_path_to_circuit_type, cache_circuits_by_build_path())
 
+/obj/machinery/post_anchor_change()
+	power_change()
+	. = ..()
```
```suggestion
	return ..()
```

no new trailing `. = ..()`

Spookerton commented on code/game/machinery/vending/security.dm:
```diff
@@ -25,7 +34,16 @@
 		/obj/item/storage/box/evidence = 6
 	)
+	rare_products = list(
+		/datum/uplink_item/item/stealth_items/sneaky_armor = 75,
```
```suggestion
		/obj/item/clothing/accessory/armor_plate/sneaky = 75,
```

Spookerton commented on code/game/machinery/vending/security.dm:
```diff
@@ -25,7 +34,16 @@
 	)
+	antag = list(
+		/obj/item/device/uplink_service/fake_command_report = 1,
+		/datum/uplink_item/item/stealth_items/sneaky_armor = 0,
```
```suggestion
		/obj/item/clothing/accessory/armor_plate/sneaky = 0,
```

Spookerton commented on code/datums/uplink/devices and tools.dm:
```diff
@@ -177,3 +177,11 @@
+	name = "Syndicate Coins"
+	item_cost = 50
+	path = /obj/item/storage/fancy/smokable/case/syndiecoins
+	desc = "A packet of five coins that unlock a secret compartment in any vending machine. \
+	Each coin unlocks on average 15 TCs worth of items. Depending on your luck, you may half or more than triple your TCs worth! \
+	These compartments cannot be accessed by hacking."
```
```suggestion
	desc = "A packet of five coins that unlock a secret compartment in any vending machine. \
		Each coin unlocks on average 15 TCs worth of items. Depending on your luck, you may half or more than triple your TCs worth! \
		These compartments cannot be accessed by hacking."
```

Spookerton commented on code/game/machinery/vending/_vending.dm:
```diff
@@ -192,20 +214,64 @@
+		if (IsShowingAntag())
+			to_chat(user, SPAN_WARNING("\The [src]'s secret compartment is already unlocked!"))
+			return TRUE
+		if (!user.unEquip(item,src))
```
```suggestion
		if (!user.unEquip(item, src))
```
since we're here

Spookerton commented on code/game/machinery/vending/wallmed1.dm:
```diff
@@ -23,8 +28,16 @@
+	antag = list(
+		/obj/item/storage/pill_bottle/assorted = 1,
+		/obj/item/storage/firstaid/sleekstab = 0
+
```
```suggestion
```

Spookerton commented on code/game/objects/items/devices/uplink_random_lists.dm:
```diff
@@ -62,6 +62,7 @@ var/global/list/uplink_random_selections_
 	items += new/datum/uplink_random_item(/datum/uplink_item/item/stealth_items/chameleon_kit)
 	items += new/datum/uplink_random_item(/datum/uplink_item/item/stealth_items/chameleon_projector)
 	items += new/datum/uplink_random_item(/datum/uplink_item/item/stealth_items/voice)
+	items += new/datum/uplink_random_item(/datum/uplink_item/item/stealth_items/cosmetic_surgery_kit, reselect_propbability = 10)
```
Is this diff pollution?

Spookerton commented on code/game/objects/items/weapons/material/coins.dm:
```diff
@@ -251,3 +204,53 @@
+	abstract_type = /obj/item/material/coin/challenge
+	applies_material_colour = FALSE
+
+
+/obj/item/material/coin/challenge/on_update_icon()
+	overlays.Cut()
+	if (isnull(string_color))
+		return
+	var/image/image = image(icon = icon, icon_state = "coin_string_overlay")
+	image.appearance_flags |= RESET_COLOR
+	image.color = string_color
+	overlays += image
+
+/obj/item/material/coin/challenge/throw_impact()
+	..()
+	if (QDELETED(src))
+		return
+	playsound(loc, 'sound/effects/coin_flip.ogg', 75, 1)
+	visible_message(
+		SPAN_WARNING("\The [src] clatters to \the [get_turf(src)]!")
+	)
+
+
+/obj/item/material/coin/challenge/verb/drop_coin()
+	set src in usr
+	set category = "Object"
+	set name = "Drop Challenge Coin"
+	var/turf/turf = usr.loc
+	if (!isturf(turf))
+		return
+	if (!usr.IsHolding(src))
+		to_chat(usr, SPAN_WARNING("You must hold \the [src] in your hands."))
+		return
+	if (!usr.unEquip(src, turf))
+		return
+	playsound(turf, 'sound/effects/coin_flip.ogg', 75, 1)
+	usr.visible_message(
+		SPAN_WARNING("\The [usr] flicks \a [src] onto the ground!"),
+		SPAN_WARNING("You flick \the [src] onto the ground!")
+	)
```
should not be moved to the bottom of the file. On the diff underneath is required.

emmanuelbassil replied:
Fixed, moved them back up. Made definition of challenge coins precede challenge coin specific verbs

Spookerton commented on code/game/objects/items/weapons/storage/uplink_kits.dm:
```diff
@@ -3,6 +3,11 @@
 	desc = "A sleek, sturdy box."
 	icon_state = "box_of_doom"
 
+/obj/item/storage/fancy/syndie_kit
+	name = "fancy box"
+	desc = "A fancy, sturdy box."
+	icon_state = "box_of_doom"
+
```
```suggestion
```
unused

Spookerton commented on code/game/objects/items/weapons/storage/fancy/_fancy.dm:
```diff
@@ -4,9 +4,11 @@
 
 	/// The item type this container cares about for behaviors like icon generation.
 	var/obj/item/key_type
+	var/obj/item/key_type_2
```
key_type should be null/text/list instead of adding duplicate fields, but that's a larger change than is easily described here. TODO later.

Spookerton commented:
In addition to the above, DNM because of sprite changes.

SierraKomodo commented:
> In addition to the above, DNM because of sprite changes.

To clarify: Expecting collisions with #33877

SierraKomodo commented on code/game/machinery/vending/security.dm:
```diff
@@ -17,6 +17,15 @@
+		If I had arms you'd be dead by now, capitalist pig.;\
+		Guilty until proven innocent!;\
+		Feeling down? We'll help you make someone else feel worse.;\
+		Civilian tested; Sol approved!;\
```
Antag slogans are split by the `;` character, so you'll need to use something else in this sentence.


```suggestion
		Civilian tested, Sol approved!;\
```

emmanuelbassil replied:
Fixed

emmanuelbassil committed:
Adds Traitor Coin

Adds some items for vendors

Slogans

More vendors!

Final antagvend

Fix spacing

Fixess

/the

Fixes

Fixes order

Merge


Auto stash before rebase of "refs/heads/vendorcoin"

emmanuelbassil commented:
Fixed merge conflicts.
Ready for review, at last.
